### PR TITLE
[jjo] juju2 bash autocompletion improvements

### PR DESCRIPTION
**Pull request by jjo**
* use python3
* --model support (including juju ssh --model <TAB> ..., etc)
* --controller support
* more aggressive caching (mainly needed for controller*model completions)

### CODE REVIEW

File: etc/bash_completion.d/juju2 [hunk 1]
```diff
@@ -1,42 +1,55 @@
+#!/bin/bash
```
Comment (bz2) [inline etc/bash_completion.d/juju2]: Lintian told someone to remove this from the file. I didn't see the message so don't know why, but not that:

```
grep -F '#!/' /usr/share/bash-completion/completions/*
```

Matches nothing on my system.

File: etc/bash_completion.d/juju2 [hunk 2]
```diff
@@ -1,42 +1,55 @@
...
 # juju-core.bash_completion.sh: dynamic bash completion for juju cmdline,
 # from parsed (and cached) juju status output.
 #
 # Author: JuanJo Ciarlante <jjo@canonical.com>
 # Copyright 2013+, Canonical Ltd.
 # License: GPLv3
 #
+# For juju2, includes --model and --controller handling:
```
Comment (bz2) [inline etc/bash_completion.d/juju2]: Per the recent mail to #juju-dev, the binary name will just be 'juju'. I was looking at this script for the packaging, and think what I need is a top level variable JUJU=juju that I can replace when installing, so the wrapping scripts 'juju-1' and 'juju-2.0' can also get correct completetion goodness.

File: etc/bash_completion.d/juju2 [hunk 3]
```diff
@@ -1,42 +1,55 @@
...
+#   juju switch <TAB>
+#   juju status --model <TAB>
+#   juju ssh --model <TAB> [... will complete with proper model's units/etc ...]
+#
 
 # Print (return) all machines
-_juju_machines_from_file() {
-python -c '
+_juju_machines_from_status_file() {
+python3 -c '
```
Comment (bz2) [inline etc/bash_completion.d/juju2]: We want this script to work on backports to trusty etc, which may have python3 available, but not installed. Can we also have a $PYTHON variable that gets set the appropriate default for the system?

File: etc/bash_completion.d/juju2 [hunk 4]
```diff
@@ -49,7 +62,36 @@ _juju_list_commands() {
...
     test -z "${1}" && return 0
-    juju help ${1} 2>/dev/null |egrep -o --  '(^|-)-[a-z-]+'|egrep -v -- '^(-m|--model)'|sort -u
+    juju help ${1} 2>/dev/null |egrep -o --  '(^|-)-[a-z-]+'|sort -u
+}
+
+# List all controllers
+_juju_list_controllers_noflags() {
+  _juju_cache_cmd 10 cat \
+    juju list-controllers --format json | \
+      python3 -c 'import json,sys; print("\n".join(json.load(sys.stdin)["controllers"].keys()))'
```
Comment (bz2) [inline etc/bash_completion.d/juju2]: Will need a space after print before ( for this to work on Python 2 and 3.

File: etc/bash_completion.d/juju2 [hunk 5]
```diff
@@ -137,9 +227,10 @@ _juju() {
...
     COMPREPLY=()
     parsing_func=$(_juju_completion_func_for_cmd "${action}" ${COMP_CWORD})
-    test -z "${parsing_func}" && return 0
-    _juju_complete_with_func "${action}" "${parsing_func}"
+    test -n "${parsing_func}" && \
+      _juju_complete_with_func "${action}" "${parsing_func}"
     return $?
 }
 complete -F _juju juju
+complete -F _juju juju2
```
Comment (bz2) [inline etc/bash_completion.d/juju2]: Same note about command name still being 'juju'.

**Comment by bz2** [comment]
Thanks!

Have added various comments inline, main note being about the binary name.

File: etc/bash_completion.d/juju [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    _juju_units_from_status_file "$@"
+    _juju_machines_from_status_file "$@"
+}
+
+# Print (return) all juju commands
+_juju_list_commands() {
+    juju help commands 2>/dev/null | awk '{print $1}'
+}
+
+# Print (return) flags for juju action, shamelessly excluding
```
Comment (jrwren) [inline etc/bash_completion.d/juju]: The comment re: shamelessly excluding -m/--model is no longer true.

Comment (jjo) [inline etc/bash_completion.d/juju]: Thanks, fixed.

File: etc/bash_completion.d/juju2 (hunk 1 above)
Comment (jjo) [inline etc/bash_completion.d/juju2]: Fixed.

File: etc/bash_completion.d/juju2 (hunk 2 above)
Comment (jjo) [inline etc/bash_completion.d/juju2]: Fixed, see comments in added "juju-version" file.

File: etc/bash_completion.d/juju2 (hunk 3 above)
Comment (jjo) [inline etc/bash_completion.d/juju2]: Fixed, added _juju_cmd_PYTHON env var.

File: etc/bash_completion.d/juju2 (hunk 4 above)
Comment (jjo) [inline etc/bash_completion.d/juju2]: Thanks, fixed.

**Comment by jjo** [comment]
I've addressed bz2 comments (tho in a different way than proposed), PTAL.

**Comment by jjo** [comment]
Appreciate another review - several improvements since last PR:
* add python2,3 loadtime detection
* 1.0 and 2.0 version handling:
- rename completion file to juju-2.0
- add "juju-version" file for run-time 1.x/2.0 completion
  (as juju-v* loaded after juju-[12]*)
- rename all functions with _juju_2_0 prefix
* cache TTL=1min (from 10min, see below)
* background completion cache refresh triggered on TTL/2:
- completion only "blocks" (waiting for actual content from eg `juju status`)
  if cache file is stale (>TTL)
- "join" on background ready implemented via reading from `coproc`d fd instead of
  `wait` for bg process to finish, which blocks bash "hard" (neither responsive to Ctrl-C)
* s/service/application/ (beta9)

**Comment by bz2** [comment]
Thanks for addressing my comments, I poked the juju team as well to get a review.

File: etc/bash_completion.d/juju-2.0 [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+    return $?
+}
+# Print (return) all machines
+_juju_2_0_machines_from_status() {
+    local cache_fname=$(_juju_2_0_juju_status_cache_fname)
+    [ -n "${cache_fname}" ] || return 0
+${_juju_cmd_PYTHON?} -c '
+import json, sys
+sys.stderr.close()
+j=json.load(sys.stdin)
```
Comment (mjs) [inline etc/bash_completion.d/juju-2.0]: spaces around = please

File: etc/bash_completion.d/juju-2.0 [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+print ("\n".join(j.get("machines", {}).keys()));
+'   < ${cache_fname}
+}
+
+# Print (return) all units, each optionally postfixed by $2 (eg. 'myservice/0:')
+_juju_2_0_units_from_status() {
+    local cache_fname=$(_juju_2_0_juju_status_cache_fname)
+    [ -n "${cache_fname}" ] || return 0
+${_juju_cmd_PYTHON?} -c '
+trail="'${2}'"
```
Comment (mjs) [inline etc/bash_completion.d/juju-2.0]: and again, and throughout

File: etc/bash_completion.d/juju-2.0 [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+import json, sys
+sys.stderr.close()
+j=json.load(sys.stdin)
+all_units=[]
+for k,v in j.get("applications", {}).items():
+    if v.get("units"):
```
Comment (mjs) [inline etc/bash_completion.d/juju-2.0]: this called also be: `all_units.extend(v.get("units", {}).keys())`

File: etc/bash_completion.d/juju-2.0 [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+    set +e
+    local model=""
+    if [[ ${COMP_LINE} =~ .*(--model|-m)\ ([^ ]+)\ (: [^ ]+\ )?.* ]];then
+       model="${BASH_REMATCH[2]}${BASH_REMATCH[3]}"
+       model="${model// /}"
+    fi
+    if [ -z "${model}" ];then
+       model=${JUJU_MODEL:-$(${_juju_cmd_JUJU_2_0?} switch)}
+    fi
+    # echo "** comp=$(set|egrep ^COMP) ** model=${model} **">&2
```
Comment (mjs) [inline etc/bash_completion.d/juju-2.0]: Should this still be here?

File: etc/bash_completion.d/juju-2.0 (hunk 9 above)
Comment (jjo) [inline etc/bash_completion.d/juju-2.0]: Thanks, changed.

File: etc/bash_completion.d/juju-2.0 (hunk 10 above)
Comment (jjo) [inline etc/bash_completion.d/juju-2.0]: Removed the commented out debugging line.

**Comment by mjs** [comment]
Thanks! This looks OK to me. 

For a future iteration of this, I wonder if it would be worth leaning on Python more to avoid the ugliness of doing fairly complex work in bash. I guess there's perhaps a speed trade off.

**Comment by jjo** [comment]
Addressed above comments, PTAL.

File: etc/bash_completion.d/juju-2.0 [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+#
+# Author: JuanJo Ciarlante <jjo@canonical.com>
+# Copyright 2016+, Canonical Ltd.
+# License: GPLv3
+#
+# Includes --model and --controller handling:
+#   juju list-models --controller <TAB>
+#   juju switch <TAB>
+#   juju status --model <TAB>
+#   juju ssh --model <TAB> [... will complete with proper model's units/etc ...]
```
Comment (mjs) [inline etc/bash_completion.d/juju-2.0]: Did you want to describe other supported completions here as well? (e.g. storage)

**Comment by mjs** [comment]
LGTM

**Comment by cherylj** [comment]
$$__JFDI__$$

**Pull request merged**